File: leetcode/p853.cpp
class Solution {
public:
    int carFleet(int target, vector<int>& position, vector<int>& speed) {
        vector<pair<double, int>> arrived_position;
        //arrived_position.assign(speed.size(), make_pair(0, -1));
        for (int i=0;i<speed.size();i++) {
            double t = double(target-position[i])/speed[i];
            arrived_position.push_back(make_pair(t, position[i]));
        }
        sort(arrived_position.begin(), arrived_position.end());
        reverse(arrived_position.begin(), arrived_position.end());
        //for (int i=0;i<arrived_position.size();i++) {
        //    printf("%f %d\n", arrived_position[i].first, arrived_position[i].second);
        //}
        int current_pos = -1;
        int total = 0;
        for (int i=0;i<arrived_position.size();i++) {
            if (arrived_position[i].second>current_pos) {
                total++;
                current_pos = arrived_position[i].second;
            }
        }
        return total;
    }
};

File: leetcode/p853.py
class Solution:
    def carFleet(self, target: int, position: List[int], speed: List[int]) -> int:
        car_heap = []
        for i in range(len(speed)):
            t = (target - position[i]) / speed[i]
            car_heap.append((-t, -1 * position[i]))
        heapq.heapify(car_heap)
        total = 0
        pos = -1
        while car_heap:
            t, p = heapq.heappop(car_heap)
            p = -1 * p
            if p < pos:
                continue
            total += 1
            pos = p
        return total
